hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+                shutdownExecutor = util::Executors::newSingleThreadExecutor(
+                        client.getName() + ".clientShutdown-");
+
+                shuffleMemberList = client.getClientProperties().getShuffleMemberList().getBoolean();
+
+                ClientConnectionManagerImpl::addressProviders = addressProviders;
+
+                connectionAttemptPeriod = networkConfig.getConnectionAttemptPeriod();
+
+                int connAttemptLimit = networkConfig.getConnectionAttemptLimit();
+                bool isAsync = client.getClientConfig().getConnectionStrategyConfig().isAsyncStart();
+
+                if (connAttemptLimit < 0) {
+                    this->connectionAttemptLimit = isAsync ? DEFAULT_CONNECTION_ATTEMPT_LIMIT_ASYNC
+                                                           : DEFAULT_CONNECTION_ATTEMPT_LIMIT_SYNC;
+                } else {
+                    this->connectionAttemptLimit = connAttemptLimit == 0 ? INT32_MAX : connAttemptLimit;
+                }
+
+            }
+
+            bool ClientConnectionManagerImpl::start() {
+                util::LockGuard guard(lock);
+                if (alive) {
+                    return true;
+                }
+                alive = true;
+
+                bool result = socketFactory.start();
+
+                socketInterceptor = client.getClientConfig().getSocketInterceptor();
+
+                if (!inSelector.start()) {
+                    return false;
+                }
+                if (!outSelector.start()) {
+                    return false;
+                }
+
+                startEventLoopGroup();
+                heartbeat.reset(new HeartbeatManager(client));
+                heartbeat->start();
+                addConnectionHeartbeatListener(boost::shared_ptr<spi::impl::ConnectionHeartbeatListener>(
+                        new spi::impl::ConnectionHeartbeatListenerDelegator(*this)));
+                connectionStrategy->start();
+
+                return result;
+            }
+
+            void ClientConnectionManagerImpl::shutdown() {
+                util::LockGuard guard(lock);
+                if (!alive) {
+                    return;
+                }
+                alive = false;
+
+                // close connections
+                BOOST_FOREACH(boost::shared_ptr<Connection> connection, activeConnections.values()) {
+                                // prevent any exceptions
+                                util::IOUtil::closeResource(connection.get(), "Hazelcast client is shutting down");
+                            }
+
+                spi::impl::ClientExecutionServiceImpl::shutdownExecutor("cluster", *clusterConnectionExecutor, logger);
+                stopEventLoopGroup();
+                connectionListeners.clear();
+                heartbeat->shutdown();
+
+                connectionStrategy->shutdown();
+
+
+                if (alive.compareAndSet(true, false)) {
```

Comment:
alive check and setting it to false is already done at the start. 
It looks like code after this point will never run. Can you check that ?

Reply:
Removed those lines.